### DISCUSSION TITLE
Versionado del proyecto

### DISCUSSION THREAD

**Discussion by laurolmer** [perealberto/decide-single-zambrano#91] [Ideas]
Buenas chicos. Abro esta discusión de cara al futuro versionado que se llevará a cabo en el proyecto. 

En primer lugar, pienso que la primera versión que debemos desplegar debe ser la 1.0.0, indicando que la situación en la que se encuentra el proyecto es una primera versión estable y que los futuros incrementos se realizarán sobre ella.

A partir de la versión 1.0.0. propongo que se siga la siguiente estructura siguiendo el versionado semántico:
x.y.z, donde: 
- "x" indica que se han añadido cambios que son incompatibles en el proyecto
- "y" indica cuando se añade una funcionalidad que sigue siendo compatible con versiones anteriores 
- "z" indica cuando se han reparado errores compatibles con versiones anteriores (parches)

¿Qué opináis?

**Comment by claugp06** [comment]
Buenas,
La verdad que veo bien la propuesta ya que se entiende facilmente las diferencias entre versiones y facilita la compresión.

**Comment by perealberto** [comment]
Hola a todos! Me parece bien la propuesta @laurolmer. Como mensaje de características de versión para la tag 1.0.0 podríamos incluir una pequeña descripción de los módulos de decide que se encuentran en el [repositorio original ](https://github.com/EGCETSII/decide). Una opción que se me ocurre, basándome en el readme.md del proyecto sería:

# Notas de versión 1.0.0

Decide está compuesto por diferentes subsistemas que son más o menos independientes entre sí y que se interconectan implementando una API concreta. Cada subsistema se encarga de una tarea concreta en el sistema de voto.

## Gateway

Para ofrecer un punto de entrada conocido para todos los subsistemas existe el llamado gateway que no es más que una ruta disponible que redirigirá todas las peticiones al subsistema correspondiente, de tal forma que cualquier cliente que use la API no tiene por qué saber en qué servidor está desplegado cada subsistema.

## Autenticación

Autenticación de votantes. En una votación debemos autenticar a los votantes y asegurarnos de que no se vota más de una vez o de que el votatne puede votar en esa votación. Para ello debemos autenticar a la persona y esta autenticación ha de ser segura, de tal forma que se reduzca la posibilidad de fraude, suplantación de identidad, etc.

## Censo

Una parte impmortante de una votación es el censo, que es el grupo de personas que pueden votar en una votación. Gestoniar el censo es una tarea importante a la hora de definir una votación y de controlar quién ha votado y quién no.

El censo está relacionado con la autenticación, pero no es lo mismo, el censo es la autorización para votar en una votación en concreto, es decir, puede estar autorizado a entrar en el sistema a través de la autenticación, pero puede que no tengas autorización para votar en una determinada votación.

## Votaciones

Una votación puede definir una o varias preguntas, con diferente número de opciones y diferentes configuraciones. Puede ser voto simple, ordenación de una lista de opciones, elección múltiple, etc.

## Cabina de votación

Interfaz para votar. Este subsistema se encarga de mostrar la interfaz de voto de una votación en concreto, permitiendo al votante votar de la forma más sencilla posible.

## Almacenamiento de votos (cifrados)

Los votos se almacenan cifrados en una base de datos, donde tenemos la relación directa de votante y voto. No se puede saber la intención de voto porque el voto estará cifrado en la base de datos, pero sí tendremos la información de quién ha votado y quién no.

## Recuento / MixNet

Susbsistema que se encarga de la parte criptográfica de una votación. La mixnet es una red de ordenadores que generarán una clave compartida para cifrar los votos. De esta manera sólo se podrá descifrar un voto cuando todas las autoridades se pongan de acuerdo, a la hora de hacer el recuento. En el proceso de recuento se desliga el votante del voto, cada autoridad aplicará un paso de barajado de votos, antes de descifrar su parte, de tal forma que cuando se obtenga el listado de votos en claro, no habrá forma de saber qué ha votado cada usuario.

## Post-procesado

Una vez realizado el recuento, tenemos una lista de números, este subsistema de encarga de traducir esa lista de números a un resultado coherente con el tipo de votación. Por ejemplo, si es una votación de tipo simple, simplemente habrá que contar el número de veces que aparece cada opción y se dará como ganadora la opción con más opciones. Este subsistema también es el encargado de aplicar diferentes reglas a los resultados, por ejemplo, aplicar reglas de paridad, ley d'Hondt, etc.

## Visualización de resultados

Este subsistema es el encargado de coger los datos obtenidos tras el post procesado y pintarlos de una manera gráfica y entendible, generando gráficas y tablas. También se encargará de generar los diferentes informes necesarios de estos resultados, ya sea en formato web, pdf, texto planto, etc.
